<a href="https://colab.research.google.com/github/NataliProhorova/Data-_Pyton-_Code_Info/blob/main/L7_Data_Exploration_CA2v4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Step 1
import libtoryes

In [ ]:

#IMPORTING LIBRARIES
import pandas as pd
import statistics as stats
import numpy as np
import matplotlib.pyplot as pl
import seaborn as sns

In [ ]:
# Importing the dataset
df = pd.read_csv('/content/spambase.csv',index_col=0))


In [ ]:
df.head()

Unnamed: 0  word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0           0            0.00               0.64           0.64           0.0   
1           1            0.21               0.28           0.50           0.0   
2           2            0.06               0.00           0.71           0.0   
3           3            0.00               0.00           0.00           0.0   
4           4            0.00               0.00           0.00           0.0   

  word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0          0.32            0.00              0.00                0.00   
1          0.14            0.28              0.21                0.07   
2          1.23            0.19              0.19                0.12   
3          0.63            0.00              0.31                0.63   
4          0.63            0.00              0.31                0.63   

   word_freq_order  ...  char_freq_;  char_freq_(  char_freq_[  char_freq_!  \
0             0.00  ...         0.00        0.000          0.0        0.778   
1             0.00  ...         0.00        0.132          0.0        0.372   
2             0.64  ...         0.01        0.143          0.0        0.276   
3             0.31  ...         0.00        0.137          0.0        0.137   
4             0.31  ...         0.00        0.135          0.0        0.135   

   char_freq_$  char_freq_#  capital_run_length_average  \
0        0.000        0.000                       3.756   
1        0.180        0.048                       5.114   
2        0.184        0.010                       9.821   
3        0.000        0.000                       3.537   
4        0.000        0.000                       3.537   

   capital_run_length_longest  capital_run_length_total  is_spam  
0                          61                       278     True  
1                         101                      1028     True  
2                         485                      2259     True  
3                          40                       191     True  
4                          40                       191     True  

[5 rows x 59 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 59 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Unnamed: 0                  4601 non-null   int64  
 1   word_freq_make              4601 non-null   float64
 2   word_freq_address           4601 non-null   float64
 3   word_freq_all               4595 non-null   float64
 4   word_freq_3d                4599 non-null   float64
 5   word_freq_our               4601 non-null   object 
 6   word_freq_over              4600 non-null   float64
 7   word_freq_remove            4601 non-null   float64
 8   word_freq_internet          4586 non-null   float64
 9   word_freq_order             4601 non-null   float64
 10  word_freq_mail              4601 non-null   float64
 11  word_freq_receive           4601 non-null   float64
 12  word_freq_will              4601 non-null   float64
 13  word_freq_people            4601 

In [ ]:
df.shape

(4601, 59)

**4601**: This number indicates the number of rows or observations in the dataset.,
**59** columns

## Dataset Overview

The dataset consists of **4601 rows** and **59 columns**.

**Data Types:**
* **Numerical:** float64(51), int64(3)
* **Categorical:** object(4)
* **bool(1)**
**Missing Values:**
* Some columns contain missing values, which will need to be addressed during data preprocessing.

**unnamed** column  appears as a default index when reading  ( CSV) file formats  It duplicates the row index, can introduce noise into machine learning models, potentially affecting their performance . ,it is only one column that I found reasonable to drop

In [ ]:
df = df.drop('Unnamed: 0', axis=1)

capital_run_length_average:

This column represents the average length of uninterrupted sequences of capital letters in an email.
Float64 is a suitable data type because averages can often be fractional numbers. For example, the average length could be 2.5.capital_run_length_longest and capital_run_length_total:

These columns represent integer counts: the length of the longest sequence of capital letters and the total number of capital letters,  that is whole numbers representing whole number counts, int64 is the correct data type.

In [ ]:
df.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

  word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0          0.32            0.00              0.00                0.00   
1          0.14            0.28              0.21                0.07   
2          1.23            0.19              0.19                0.12   
3          0.63            0.00              0.31                0.63   
4          0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  is_spam  
0                       278     True  
1                      1028     True  
2                      2259     True  
3                       191     True  
4                       191     True  

[5 rows x 58 columns]

irst, we need to address the non-numeric columns. Columns like word_freq_our and word_freq_000 are likely incorrectly labeled as object dtype. These columns should ideally be numerical.Ensuring Values are Between 0 and 1
If the values in your columns are already within the range of 0 to 1, you don't need to explicitly clip them. However, if there are values outside this range, you can use:

Python
df['word_freq_make'] = df['word_freq_make'].clip(lower=0, upper=1)    Formatting to 2 Decimal Places
df['word_freq_make'] = df['word_freq_make'].round(2)
Handling Missing Values
You'll need to address missing values (NaN) in your dataset. You can choose from several strategies:

Imputation: Replace missing values with a suitable value (e.g., mean, median, mode, or a predictive model).
Deletion: Remove rows or columns with missing values.
Model-Based Imputation: Use machine learning models to predict missing values.
# Assuming your DataFrame is named 'df'

# Convert non-numeric columns to numeric
df['word_freq_our'] = pd.to_numeric(df['word_freq_our'], errors='coerce')
# ... similar for other non-numeric columns

# Clip values to 0-1 range
df = df.clip(lower=0, upper=1)

# Round values to 2 decimal places
df = df.round(2)

# Handle missing values (e.g., imputation)
df.fillna(df.mean(), inplace=True)  # Replace missing values with column mean

# Formatting Numbers for Machine Learning By standardizing the number of decimal places
with word and character frequencies, two decimal places
 will be more reasonable. This level of precision improve  unnecessary complexity, avoid potential issues, better Model Performance and Storage Efficiency .

In [ ]:
#after standartisation of all fetures to numerical format
df = df.round(2)

In [ ]:
missing_values = df.isnull().sum()
print(missing_values)

word_freq_make                  0
word_freq_address               0
word_freq_all                   6
word_freq_3d                    2
word_freq_our                   0
word_freq_over                  1
word_freq_remove                0
word_freq_internet             15
word_freq_order                 0
word_freq_mail                  0
word_freq_receive               0
word_freq_will                  0
word_freq_people                0
word_freq_report                0
word_freq_addresses             9
word_freq_free                  0
word_freq_business              0
word_freq_email                 0
word_freq_you                   0
word_freq_credit                0
word_freq_your                  0
word_freq_font                  0
word_freq_000                   0
word_freq_money                 0
word_freq_hp                    0
word_freq_hpl                   0
word_freq_george                0
word_freq_650                   0
word_freq_lab                   0
word_freq_labs

In [ ]:
df.fillna(0, inplace=True)

In [ ]:
import pandas as pd

df.bfill(inplace=True)


In [ ]:
# Select numerical features
numerical_features = df.select_dtypes(include=['float64', 'int64']).columns

for feature in numerical_features:
    df[feature] = pd.to_numeric(df[feature], errors='coerce')
# Check for missing values in numerical_features

missing_values = df[feature].isnull().sum()
print(missing_values)


0


In [ ]:
df.describe()


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
count     4601.000000        4601.000000    4601.000000   4601.000000   
mean         0.104553           0.213015       0.280656      0.065425   
std          0.305358           1.290575       0.504143      1.395151   
min          0.000000           0.000000       0.000000      0.000000   
25%          0.000000           0.000000       0.000000      0.000000   
50%          0.000000           0.000000       0.000000      0.000000   
75%          0.000000           0.000000       0.420000      0.000000   
max          4.540000          14.280000       5.100000     42.810000   

       word_freq_over  word_freq_remove  word_freq_internet  word_freq_order  \
count     4601.000000       4601.000000         4601.000000      4601.000000   
mean         0.095901          0.114208            0.105295         0.090067   
std          0.273824          0.391441            0.401071         0.278616   
min          0.000000          0.000000            0.000000         0.000000   
25%          0.000000          0.000000            0.000000         0.000000   
50%          0.000000          0.000000            0.000000         0.000000   
75%          0.000000          0.000000            0.000000         0.000000   
max          5.880000          7.270000           11.110000         5.260000   

       word_freq_mail  word_freq_receive  ...  word_freq_conference  \
count     4601.000000        4601.000000  ...           4601.000000   
mean         0.239413           0.059824  ...              0.031869   
std          0.644755           0.201545  ...              0.285735   
min          0.000000           0.000000  ...              0.000000   
25%          0.000000           0.000000  ...              0.000000   
50%          0.000000           0.000000  ...              0.000000   
75%          0.160000           0.000000  ...              0.000000   
max         18.180000           2.610000  ...             10.000000   

       char_freq_;  char_freq_(  char_freq_[  char_freq_!  char_freq_$  \
count  4601.000000  4601.000000  4601.000000  4601.000000  4601.000000   
mean      0.038574     0.138978     0.016962     0.269009     0.075766   
std       0.243410     0.270336     0.109387     0.815661     0.245808   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.060000     0.000000     0.000000     0.000000   
75%       0.000000     0.190000     0.000000     0.320000     0.050000   
max       4.380000     9.750000     4.080000    32.480000     6.000000   

       char_freq_#  capital_run_length_average  capital_run_length_longest  \
count  4601.000000                 4601.000000                 4601.000000   
mean      0.044208                    5.191519                   52.172789   
std       0.429387                   31.729457                  194.891310   
min       0.000000                    1.000000                    1.000000   
25%       0.000000                    1.590000                    6.000000   
50%       0.000000                    2.280000                   15.000000   
75%       0.000000                    3.710000                   43.000000   
max      19.830000                 1102.500000                 9989.000000   

       capital_run_length_total  
count               4601.000000  
mean                 283.289285  
std                  606.347851  
min                    1.000000  
25%                   35.000000  
50%                   95.000000  
75%                  266.000000  
max                15841.000000  

[8 rows x 53 columns]

In [ ]:
# Get the descriptive statistics
description = df.describe() # Make sure df is your DataFrame

# Extract the 'max' row and convert it to a Series
max_values = description.loc['max']

# Sort the Series in descending order
sorted_max_values = max_values.sort_values(ascending=False)

# Print the sorted maximum values
print(sorted_max_values)

capital_run_length_total      15841.00
capital_run_length_longest     9989.00
capital_run_length_average     1102.50
word_freq_3d                     42.81
word_freq_george                 33.33
char_freq_!                      32.48
word_freq_edu                    22.05
word_freq_re                     21.42
word_freq_hp                     20.83
word_freq_85                     20.00
word_freq_free                   20.00
word_freq_project                20.00
char_freq_#                      19.83
word_freq_you                    18.75
word_freq_data                   18.18
word_freq_mail                   18.18
word_freq_credit                 18.18
word_freq_font                   17.10
word_freq_meeting                14.28
word_freq_address                14.28
word_freq_lab                    14.28
word_freq_telnet                 12.50
word_freq_money                  12.50
word_freq_your                   11.11
word_freq_internet               11.11
word_freq_pm             

In [ ]:
print(df['is_spam'].head())

0    True
1    True
2    True
3    True
4    True
Name: is_spam, dtype: bool


In [ ]:
# Assuming you have your DataFrame 'df'

# Move 'is_spam' column to the first position (if needed)
cols = ['is_spam'] + [col for col in df.columns if col != 'is_spam']
df = df[cols]

In [ ]:
df.head()

is_spam  word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0     True            0.00               0.64           0.64           0.0   
1     True            0.21               0.28           0.50           0.0   
2     True            0.06               0.00           0.71           0.0   
3     True            0.00               0.00           0.00           0.0   
4     True            0.00               0.00           0.00           0.0   

  word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0          0.32            0.00              0.00                0.00   
1          0.14            0.28              0.21                0.07   
2          1.23            0.19              0.19                0.12   
3          0.63            0.00              0.31                0.63   
4          0.63            0.00              0.31                0.63   

   word_freq_order  ...  char_freq_!  char_freq_$  char_freq_#  \
0             0.00  ...         0.78         0.00         0.00   
1             0.00  ...         0.37         0.18         0.05   
2             0.64  ...         0.28         0.18         0.01   
3             0.31  ...         0.14         0.00         0.00   
4             0.31  ...         0.14         0.00         0.00   

   capital_run_length_average  capital_run_length_longest  \
0                        3.76                          61   
1                        5.11                         101   
2                        9.82                         485   
3                        3.54                          40   
4                        3.54                          40   

   capital_run_length_total  is_spam_true  is_spam_false  is_spam  is_spam  
0                       278             1              0     True    False  
1                      1028             1              0     True    False  
2                      2259             1              0     True    False  
3                       191             1              0     True    False  
4                       191             1              0     True    False  

[5 rows x 62 columns]

In [ ]:
new_columns = [
    (df['is_spam'] == True).astype(int),  # is_spam_true
    (df['is_spam'] == False).astype(int)  # is_spam_false
]
df = pd.concat([df] + new_columns, axis=1, names=['is_spam'] + ['is_spam_true', 'is_spam_false'])

# Print the head of the DataFrame to see the new columns
print(df[['is_spam', 'is_spam_true', 'is_spam_false']].head())



   is_spam  is_spam  is_spam  is_spam_true  is_spam_false
0     True        1        0             1              0
1     True        1        0             1              0
2     True        1        0             1              0
3     True        1        0             1              0
4     True        1        0             1              0


In [ ]:
q1

In [ ]:


df['capital_run_length_total'] = pd.to_numeric(df['capital_run_length_total'], errors='coerce')
df['is_spam'] = df['is_spam'].astype(bool)
correlation = df['capital_run_length_total'].corr(df['is_spam'])
print(f"Correlation between capital run length and spam: {correlation:.2f}")

In [ ]:
# Count the number of True and False values
# Calculate the total number of rows
total_rows = len(df)
true_count = df['is_spam'].sum()
false_count = total_rows - true_count

# Calculate percentages
true_percentage = (true_count / total_rows) * 100
false_percentage = (false_count / total_rows) * 100
print(f"Percentage of True values: {true_percentage.values[0]:.2f}%")
print(f"Percentage of False values: {false_percentage.values[0]:.2f}%")

Percentage of True values: 39.40%
Percentage of False values: 60.60%


In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd

# Your provided data (create a DataFrame)
data = {
    'capital_run_length_total':df['capital_run_length_total']
    'is_spam':df['is_spam']
    'is_spam_true': df['is_spam_true']
    'is_spam_false':df['is_spam_false']
}
df = pd.DataFrame(data)

# Create bar plot
plt.figure(figsize=(10, 6))
sns.barplot(x=df.index, y='capital_run_length_total', hue='is_spam', data=df, palette={True: 'red', False: 'green'})
plt.xlabel('Spam Frequency')  # Changed x-axis label
plt.ylabel('Capital Run Length Total')
plt.title('Capital Run Length Total by Spam Classification')
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.show()


In [ ]:
nhimport seaborn as sns
import matplotlib.pyplot as plt

df = df.select_dtypes(include=['number'])

# Calculate the correlation matrix
corr_matrix = df.corr()

# Create a heatmap
plt.figure(figsize=(12, 10))
sns.heatmap(corr_matrix, annot=True, cmap='coolwarm')

plt.title('Correlation Matrix')
plt.show()

In [ ]:
# Assuming 'is_spam' is a binary column (0 or 1)
sns.barplot(x='capital_run_length_total', y='word_freq_make', data=df, hue='capital_run_length_total')
plt.xlabel('Spam/Not Spam')
plt.ylabel('Word Frequency of "make"')
plt.title('Word Frequency of "make" by Spam/Not Spam')
plt.show()

In [ ]:
df = df = df.select_dtypes(include=['number'])

# Scatter plot
plt.figure(figsize=(8, 6))
plt.scatter(df['word_freq_address'], df['char_freq_$'])
plt.xlabel('Word Frequency of "address"')
plt.ylabel('Character Frequency of "$"')
plt.title('Relationship Between Word Frequency and Character Frequency')
plt.show()

 pivoting the dataset on the is_spam column,to identify spam-specific keywords  combinations ,character sequences  and punctuation usage between two categories by calculate the mean value of word_freq_() for both spam and non-spam emails

In [ ]:
# #calculate the mean value of word_freq_address for both spam and non-spam emails for  `df` with columns `is_spam` and `word_freq_address`
pivoted_df = df.pivot_table(values='capital_run_length_total', index='is_spam', aggfunc='mean')
print(pivoted_df)

ValueError: Grouper for 'is_spam' not 1-dimensional

In [ ]:
# Pivot the data, calculating the mean for each word frequency by spam category
pivoted_df = df.pivot_table(
    values=['word_freq_address', 'word_freq_make', 'word_freq_address', 'word_freq_all', 'word_freq_3d',
            'word_freq_our', 'word_freq_over', 'word_freq_remove',
            'word_freq_internet', 'word_freq_order', 'word_freq_mail',
            'word_freq_receive', 'word_freq_will', 'word_freq_people',
            'word_freq_report', 'word_freq_addresses', 'word_freq_free',
            'word_freq_business', 'word_freq_email', 'word_freq_you',
            'word_freq_credit', 'word_freq_your', 'word_freq_font', 'word_freq_000',
            'word_freq_money', 'word_freq_hp', 'word_freq_hpl', 'word_freq_george',
            'word_freq_650', 'word_freq_lab', 'word_freq_labs', 'word_freq_telnet',
            'word_freq_857', 'word_freq_data', 'word_freq_415', 'word_freq_85',
            'word_freq_technology', 'word_freq_1999', 'word_freq_parts',
            'word_freq_pm', 'word_freq_direct', 'word_freq_cs', 'word_freq_meeting',
            'word_freq_original', 'word_freq_project', 'word_freq_re',
            'word_freq_edu', 'word_freq_table', 'word_freq_conference'],index='is_spam',aggfunc='mean')
print(pivoted_df)

In [ ]:

# Assuming you have the pivoted DataFrame 'pivoted_df'

# Melt the DataFrame
melted_df = pivoted_df.reset_index().melt(
    id_vars='is_spam',
    var_name='word_freq_type',
    value_name='value'
)

print(melted_df.head(3))

In [ ]:

# Filter for spam emails+word freq
spam_df = melted_df[melted_df['is_spam'] == True]

# Sort by word_freq_type
sorted_spam_df = spam_df.sort_values(by='value', ascending=False)

# Print the result
print(sorted_spam_df)


In [ ]:
spam_df.info()

In [ ]:
print(spam_df.head())

In [ ]:
# Filter for spam emails+word freq
not_spam_df = melted_df[melted_df['is_spam'] == False]

# Sort by word_freq_type
sorted_not_spam_df =not_spam_df.sort_values(by='value', ascending=False)

# Print the result
print(sorted_not_spam_df)

In [ ]:
not_spam_df.info()

combine two DataFrames, spam_df and non_spam_df, into a single DataFrame where:
The word_freq_type column is the index.
There are two columns: True (for spam) and False (for non-spam), representing the average word frequency for that word type in each category.
Code Implementation:

In [ ]:
print(spam_df.columns)  # Print column names of spam_df
print(not_spam_df.columns)  # Print column names of not_spam_df

In [ ]:

# Handle missing values (if necessary)
# Handle missing values (if necessary)
sorted_spam_df.fillna(0, inplace=True)  # Replace missing values with 0 (or a suitable value)
sorted_not_spam_df.fillna(0, inplace=True)

In [ ]:
print(len(sorted_spam_df))
print(len(sorted_not_spam_df))

In [ ]:
# Merge the DataFrames (optional outer join)
combined_df = sorted_spam_df.merge(sorted_not_spam_df, on='word_freq_type', how='outer')

print(combined_df)

In [ ]:
combined_df = combined_df.set_index('word_freq_type')


In [ ]:
print(combined_df)

In [ ]:

combined_df.rename(columns={'is_spam_x': 'Spam', 'is_spam_y': 'Not Spam'},inplace=True)

In [ ]:
import matplotlib.pyplot as plt

# Assuming the pivoted DataFrame is named 'pivoted_df'

# Visualize the top 10 words for spam and non-spam
top_10_spam = pivoted_df.sort_values(by=True, ascending=False).head(10)
top_10_non_spam = pivoted_df.sort_values(by=False, ascending=False).head(10)

plt.figure(figsize=(12, 6))
plt.subplot(1, 2, 1)
top_10_spam.plot(kind='bar', title='Top 10 Words in Spam')

plt.subplot(1, 2, 2)
top_10_non_spam.plot(kind='bar', title='Top 10 Words in Non-Spam')

plt.show()

In [ ]:
import pandas as pd
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler


In [ ]:
# Standardize the numerical features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(df[numerical_features])


# Perform PCA PCA  transforms the data into a new set of features called principal components (PCs).  These PCs capture the most significant variations in the original data.
pca = PCA()
pca.fit(X_scaled)


In [ ]:
#print a list of values representing the proportion of variance explained by each principal component.
explained_variance_ratio = pca.explained_variance_ratio_
print(explained_variance_ratio)

In [ ]:
# calculates and prints the total number of principal components generated by PCA.
n_components = len(explained_variance_ratio)
print(f"Number of Components: {n_components}")

In [ ]:
principal_components = pca.components_
print(principal_components.shape)  # Prints the dimensions of the components matrix

In [ ]:
# Apply PCA to reduce dimensionality
pca = PCA(n_components=n_components)
X_pca = pca.fit_transform(X_scaled)



# Create a new DataFrame with the reduced features
df_pca = pd.DataFrame(X_pca, columns=[f'PC{i+1}' for i in range(n_components)])

df_pca.head()